blog-posts/event-sourcing-serveless/event-sourcing-serveless.md

```diff
@@ -0,0 +1,166 @@
+
+![Logo kumo](./assets/ArchiMissingDataSynchro.png 'Logo Kumo')
+
+This architecture is an implementation of **CQRS interface** for the use case of the bank which has to store all credit and debit events. There are two **commands** writing credit and debit events in an **event ledger**. There are two **queries** that fetch each its associated **projection**, the total balance of the account and the last transaction made.
+
+As you can see, the missing part of the diagram is the synchronization between the write model and the read model. That's the block we will focus on.
+
+## First and foremost, in event sourcing, the order of the events is essential
+
+All credit and debit events for every user are stored in the event ledger. The access pattern I chose is the `userId` as the **partition key** and the number of the event also called the `version` as the **sort key**.
+
+![Logo kumo](./assets/event-store-schema.png 'Logo Kumo')
+
+> :question: Why using a `version` field when we can simply take the `timestamp` to keep track of the order of the events?
+
+Two commands coming in **concurrently** will write data based on the **same version of the aggregate** because they fetched it simultaneously. Let's say a user has 50$ on his bank account and two purchases come in at the same time, one of 40$ and one of 25$. Both event would **both succeed the business validation tests**. The composite key `userId`/`timestamp` of both event would be unique because timestamps would be really close but not exactly the same. Hence, both events would be written in the event ledger and you would be **overdrawn** which is **forbidden** in our bank. On the other hand, if the **sort key** is the `version` of the event, **one out of both would be invalid**. Exact simultaneity doesn't exist in computer science, so the first purchase of 40$ would be written and then shortly after the second purchase would be rejected because the event to be written is based on the same version of the last event.
+
+## How is the synchronization between read and write models ensured ?
+
+> The pain point in CQRS is to have **reliable reading data** compared to the source of truth. Projections need to be updated according to events written in the event ledger.
+
+### A fanout listening to Dynamodb Streams emits the events
+
+The objective here is to **react in real time to events** written inside the event ledger in order to compute the projections. For that a lambda as a **fanout** can be used by plugging it to the **DynamoDB streams** of the **event ledger**. `INSERT` actions have to be filtered with the filter patterns so that only theses actions trigger the fanout. DynamoDB batches its streams of events in arrays, so for each of them an event is published in Amazon EventBridge. These EventBridge events are those triggering the lambda that project on the read models.
```

Comment:
Idem, je rajouterais un petit schéma d'archi

Reply:
Du plus, l'utilisation du présent (en général) est un peu étrange. Tu te projette dans une appli bancaire, donc utilise plutôt des `we could do this`, `we could do that`

Reply:
Je comprends pas trop ce que tu veux dire par là.
Je montre justement ce qu'on a mis en place, il n'y a pas tellement de débat.
Et puis je trouve que ça fait transparaitre un manque de confiance en ce qu'on a implémenté si je me mettais à parler au conditionnel tout au long de l'article.

Reply:
Yes un schéma ça peut rendre cette aprtie plus claire 👍